pkg/kubelet/allocation/allocation_manager.go

```diff
@@ -114,6 +133,82 @@ func (m *manager) UpdatePodFromAllocation(pod *v1.Pod) (*v1.Pod, bool) {
 	return updatePodFromAllocation(pod, allocs)
 }
 
+func GetPodResizeAllocation(pod *v1.Pod) (*PodResourceSummary, bool) {
+	resizeManager.mutex.RLock()
+	defer resizeManager.mutex.RUnlock()
+
+	resizeAllocation, exists := resizeManager.podResizeAllocation[pod.UID]
+	if !exists {
+		return nil, false
+	}
+	return resizeAllocation, true
+}
+
+func ClearPodResizeAllocation(pod *v1.Pod) {
+	resizeManager.mutex.Lock()
+	defer resizeManager.mutex.Unlock()
+	delete(resizeManager.podResizeAllocation, pod.UID)
+}
+
+// Compare reqOld and reqNew, if different, store reqNew resource to diffReq
+func compareAndSaveDiff(reqOld, reqNew v1.ResourceRequirements) v1.ResourceRequirements {
+	var diffReq v1.ResourceRequirements
+
+	// Compare Requests resource
+	diffReq.Requests = make(v1.ResourceList)
+	for key, reqNewVal := range reqNew.Requests {
+		if reqOldVal, exists := reqOld.Requests[key]; !exists || !reqOldVal.Equal(reqNewVal) {
+			diffReq.Requests[key] = reqNewVal
+		}
+	}
+
+	// Compare Limits resource
+	diffReq.Limits = make(v1.ResourceList)
+	for key, reqNewVal := range reqNew.Limits {
+		if reqOldVal, exists := reqOld.Limits[key]; !exists || !reqOldVal.Equal(reqNewVal) {
+			diffReq.Limits[key] = reqNewVal
+		}
+	}
+	return diffReq
+}
+
+// Check Requests and Limists are empty
+func isResizeResourceEmpty(req v1.ResourceRequirements) bool {
+	return len(req.Requests) == 0 && len(req.Limits) == 0
+}
+
+// Compare and save Resize resource to PodResizeReqs
+func updatePodResizeAllocation(pod *v1.Pod, c v1.Container, req1, req2 v1.ResourceRequirements) {
+
+	diffReq := compareAndSaveDiff(req1, req2)
+
+	if isResizeResourceEmpty(diffReq) {
+		return
+	}
+
+	podResizeAlloc, exists := GetPodResizeAllocation(pod)
+	if !exists {
+		podResizeAlloc = &PodResourceSummary{}
+	}
+
+	resizeManager.mutex.Lock()
+	defer resizeManager.mutex.Unlock()
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.SidecarContainers) && podutil.IsRestartableInitContainer(&c) {
+		if podResizeAlloc.InitContainers == nil {
+			podResizeAlloc.InitContainers = make(map[string]v1.ResourceRequirements)
+		}
+		podResizeAlloc.InitContainers[c.Name] = diffReq
+	} else {
+		if podResizeAlloc.Containers == nil {
```

Comment:
Is this true that if` features.SidecarContainers` is not enabled then we assume that c is not an initcontainer? This doesn't look entirely correct to me.

Reply:
Well, this logic makes sense to me. However, if  `features.SidecarContainers` is not enabled or `IsRestartableInitContainer` is not enabled we end up here and unconditionally update `podResizeAlloc.Containers`, which is what I don't understand. 
In other words, if c is an initcontainer, but `features.SidecarContainers` is not enabled this code would update podResizeAlloc.Containers, which looks incorrect.